go/login/server/server.go

```diff
@@ -0,0 +1,299 @@
+}
+
+// NewServer creates a new server with provided Authorizer and, optionally, selected options
+func NewLoginServer(a Authorizer, options ...func(*LoginServer) error) (*LoginServer, error) {
+	srv := LoginServer{
+		auth:        a,
+		Keys:        NewKeyManager(),
+		responseLen: MaxResponseSize,
+	}
+	srv.router = srv.newRouter()
+
+	for _, option := range options {
+		if err := option(&srv); err != nil {
+			return nil, err
+		}
+	}
+
+	return &srv, nil
+}
+
+// ResponseLen is an option to be provided to NewServer on creation. Its sets response len to
+// n characters long response.
+func ResponseLen(n uint8) func(srv *LoginServer) error {
+	return func(srv *LoginServer) error {
+		if 1 > n || MaxResponseSize < n {
+			return ErrInvalidResponseLen
+		}
+		srv.responseLen = n
+		return nil
+	}
+}
+
+// UserHeader is an option to be provided to NewServer on creation. Its sets UserHeader to s.
+func UserHeader(s string) func(srv *LoginServer) error {
+	return func(srv *LoginServer) error {
+		srv.userHeader = s
+		return nil
+	}
+}
+
+//----- code onwards will be deprecated as soon as integration with go-glome-login is done -----
+func (s *LoginServer) newRouter() *mux.Router { // Review Error Message
+	r := mux.NewRouter().UseEncodedPath()
+
+	handler := func(w http.ResponseWriter, r *http.Request) {
+		params := mux.Vars(r)
+
+		user := r.Header.Get(s.userHeader)
+
+		if err := handleVersion(params["V"]); err != nil {
+			log.Printf(err.Error())
+			fmt.Fprintln(w, err.Error())
+			return
+		}
+
+		dialog, tag, err := s.handleHandshakeV1(params["handshake"])
+		if err != nil {
+			log.Printf(err.Error())
+			fmt.Fprintln(w, err.Error())
+			return
+		}
+
+		hostID, err := readHost(params["hostID"])
+		if err != nil {
+			log.Printf(err.Error())
+			fmt.Fprintln(w, err.Error())
+			return
+		}
+
+		action, err := readAction(params)
+		if err != nil {
+			log.Printf(err.Error())
+			fmt.Fprintln(w, err.Error())
+			return
+		}
+
+		msg, err := readMessage(hostID, action)
+		if err != nil {
+			log.Printf(err.Error())
+			fmt.Fprintln(w, err.Error())
+			return
+		}
+
+		if err = verifyReceivedTag(dialog, tag, msg); err != nil {
+			log.Printf(err.Error())
+			fmt.Fprintln(w, err.Error())
+			return
+		}
+
+		hostIDType := ""
+		s.printToken(w, dialog, hostID, hostIDType, action, msg, user)
+	}
+
+	r.Path("/v{V:[0-9]+}/{handshake}/{hostID}/").HandlerFunc(handler)
+	r.Path("/v{V:[0-9]+}/{handshake}/{hostID}/{action:[a-zA-Z0-9+/=%]+}/").
+		HandlerFunc(handler)
+
+	return r
+}
+
+func handleVersion(V string) error {
+	switch i, err := strconv.Atoi(V); {
+	case err != nil:
+		return err
+	case i != 1:
+		return ErrVersionNotSupported
+	}
+	return nil
+}
+
+func (s *LoginServer) handleHandshakeV1(h string) (*glome.Dialog, []byte, error) {
+	data, err := base64.URLEncoding.DecodeString(h)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if len(data) < 32 { //handshake must have 66 bytes (2 prefixtype+prefix7, 32 eph, 32 prefixN)
```

Comment:
What is meant by 32 and how was it got? Is it suitable to make a constant from it?